### DISCUSSION TITLE
How to read ignoreregex from file.

### DISCUSSION THREAD

**Discussion by W1T3C** [fail2ban/fail2ban#3623] [Q&A] [answered]
Hello,

My goal is to separate 'ignoreregex' into a separate file to preserve 'ignoreregex' after bulk updates of filters on my web servers. How to do that? 

```ini
[INCLUDES]

[Definition]

failregex = (?i)([a-zA-Z0-9.-]+)\.([a-zA-Z]{2,4}) <HOST> .*upstream=- upstream_cache_status=- upstream_response_time=-.*$
        (?i)([a-zA-Z0-9.-]+)\.([a-zA-Z]{2,4}) <HOST> .*upstream_cache_status=MISS.*$

ignoreregex = /path/to/file/with/ignoreregex 
```

**Comment by sebres** [comment, marked as answer]
> My goal is to separate 'ignoreregex' into a separate file to preserve 'ignoreregex' after bulk updates of filters on my web servers

The approach looks wrong to me:
1. you must not change stock filters at all, if you want some changes it is possible via local config file - create `*.local` (that override settings of `*.conf`) and set only parameters you want to overwrite. So in your case it'd look like:
```ini
# /etc/filter.d/nginx-auth.local
[Definition]
ignoreregex = ^\s*this message must be ignored
```
2. Normally `ignoreregex` is not needed at all (why you need to ignore something that matches `failregex`, better would be to change `failregex` in order that would not match unexpected message). For instance you can use negative lookahead and lookbehind in RE or rewrite `failregex` more precise.

If the intention is to ignore specific host, maybe you should rather look at `ignoreip` or `ignorecommand`.

**Reply by W1T3C** [reply]
Thanks for the fast response and your time. I'm afraid I was not precise enough, and I apologize for that. I have 15 web servers. When I make changes in the fail2ban configuration, I prepare a new release of the fail2ban package with new versions of filters and jails, and deploy them to the web servers using our own repository. However, on some web servers, I need to be able to preserve exceptions in the form of 'ignoreregex' specific to that server. I do not change the stock filters, but one of my own filters.

For example:
[root@web15 ~]# cat /etc/fail2ban/filter.d/apache-ddos.conf
[INCLUDES]

[Definition]

failregex = (?i)([a-zA-Z0-9.-]+)\.([a-zA-Z]{2,4}) <HOST> .*upstream=- upstream_cache_status=- upstream_response_time=-.*$
        (?i)([a-zA-Z0-9.-]+)\.([a-zA-Z]{2,4}) <HOST> .*upstream_cache_status=MISS.*$

ignoreregex = /etc/fail2ban/filter.d/apache-ddos-ignoreregex.conf


If I change failregex / add new failregex in /etc/fail2ban/filter.d/apache-ddos.conf , I would like to my 'ignoreregex' be preserved.
Question is - can ignoreregex read a file ?

**Reply by sebres** [reply]
> Question is - can ignoreregex read a file ?

Exactly that - No. 
But as already said you could either put your `ignoreregex` it into `filter.d/apache-ddos.local` as described above (it would then overwrite `ignoreregex` specified in `filter.d/apache-ddos.conf`).
Or add a new entry to section `[INCLUDE]` with `after = /path/to/your-file-with-ignore.conf` and put there `ignoreregex` (in definition section).

But again, as already said too, it is not good solution at all (however it depends, but...), normally even the exceptions should not be found by `failregex`, so the correct way may be not to put `failregex` matching them, or to overwrite it with some more precise `failregex` or use some negative lookahead inside. 
That can be also put to some params and included into local (or some include file), for example:

### filter.d/some-filter.conf (same on all servers)
```ini
[Definition]
lookahead = 
failregex = ^some failure from <lookahead>[^\[]*\[<ADDR>\]
```

### filter.d/some-filter.local (1st server)
empty or doesn't exists

### filter.d/some-filter.local (2nd server)
```ini
[Definition]
# exception (ignore "from unknown"):
lookahead = (?!unknown\b)
```

### filter.d/some-filter.local (3rd server)
```ini
[Definition]
# only match "from unknown", ignore others:
lookahead = (?=unknown\b)
```

<table>
<thead><tr><th>message</th><th>server</th><th>matches</th></tr></thead>
<tbody>
<tr><td>some failure from [192.0.2.1]</td><td rowspan="3">1st</td><td>+</td></tr>
<tr><td>some failure from unknown [192.0.2.2]</td><td>+</td></tr>
<tr><td>some failure from known [192.0.2.3]</td><td>+</td></tr>
<tr><td>some failure from [192.0.2.1]</td><td rowspan="3">2nd</td><td>+</td></tr>
<tr><td>some failure from unknown [192.0.2.2]</td><td><sub>(not unknown)</sub></td></tr>
<tr><td>some failure from known [192.0.2.3]</td><td>+</td></tr>
<tr><td>some failure from [192.0.2.1]</td><td rowspan="3">3rd</td><td><sub>(only unknown)</sub></td></tr>
<tr><td>some failure from unknown [192.0.2.2]</td><td>+</td></tr>
<tr><td>some failure from known [192.0.2.3]</td><td><sub>(only unknown)</sub></td></tr>
</tbody>
</table>

By the way your regex are vulnerable a bit (not anchored and contains catch-alls), they may be affected by some inject on foreign input and are probably very slow.
